### ISSUE TITLE
Random 3000 ms delays behind nginx reverse proxy

### ISSUE THREAD

**Issue by 4oo4** [gotify/server#288]
**Have you read the documentation?**
- [x] Yes, but it does not include related information regarding my question.
- [ ] Yes, but the steps described in the documentation do not work on my machine.
- [ ] Yes, but I am having difficulty understanding it and wants clarification.

**You are setting up gotify in**
- [ ] Docker
- [x] Linux native platform
- [ ] Windows native platform

**Describe your problem**
I'm having issues where randomly loading the WebUI, or sending a push via gotify-cli will result in a 3000 ms delay. When loading these from the backend URLs, my response times are pretty much always excellent (< 100 ms). I monitory my instance and that delay seems to happen on a semi-regular interval (I check my gotify URL about once every 60 seconds). If I call try to load a gotify URL a few times in a row while testing, I would say it happens about once every 3 attempts.

I cannot reproduce the issue by bypassing the nginx reverse proxy, so the issue appears to be something in the reverse proxy config. I have a number of different webapps I run behind this reverse proxy (all using very similar nginx configs), gotify is the only one that exhibits this behavior.

I did a few things to see what might be wrong with my config, but couldn't seem to come up with anything.
 * I saw that there was a large (3.1 MB) chunk.js.map file, sometimes the browser logs would show that as the blocking request, so I wondered if this was because I was using HTTP/2 in my config (even though I think HTTP/2 was designed to prevent this kind of blocking?). However this still happened after disabling HTTP/2, and sometimes it would also occur on the request to `/version`.
 * Using memcached in my config to cache assets. I tried disabling memcached, but still reproduced the issue. I was concerned that memcached might be trying to cache the websocket requests too, but from what I can tell memcached ignores those, but the fact that it occurs without memcached means this isn't the issue.
  * Disabling gzip, no change.
  * Tweaking proxy settings (disabling `proxy_buffering`, setting `proxy_socket_keepalive`, etc.)

**Has anyone else experienced an issue like this, or can someone see what I'm doing wrong in my nginx configs?**

**Any errors, logs, or other information that might help us identify your problem**

<details><summary>Browser network logs</summary><br><pre>

![Selection_008](https://user-images.githubusercontent.com/15698835/77854025-32a80180-71d7-11ea-89eb-a10c5052cbaa.png)

</pre></details>

<details><summary>gotify-cli testing</summary><br>

```
$ time gotify push test > /dev/null
real	0m0.114s
user	0m0.096s
sys	0m0.024s

$ time gotify push test > /dev/null
real	0m0.106s
user	0m0.112s
sys	0m0.008s

$ time gotify push test > /dev/null
real	0m3.108s
user	0m0.112s
sys	0m0.012s
```
</details>

<details><summary>Monitoring - Gotify Response Times</summary><br><pre>

![Selection_002](https://user-images.githubusercontent.com/15698835/77854323-23c24e80-71d9-11ea-9b05-a6b91f5c83ae.png)

</pre></details>

<details><summary>Nginx configs</summary><br>

```
user www-data;
worker_processes 4;
pid /run/nginx.pid;

events {
	worker_connections 1024;
	multi_accept on;
}

http {

	sendfile on;
	tcp_nopush on;
	tcp_nodelay on;
	keepalive_timeout 1800;
	types_hash_max_size 2048;
	server_tokens off;
	etag off;
	port_in_redirect off;

	server_names_hash_max_size 1024;
	server_names_hash_bucket_size 1024;
	server_name_in_redirect off;
	proxy_headers_hash_max_size 2048;
	proxy_headers_hash_bucket_size 2048;


	gzip on;
	gzip_disable "msie6";
	gzip_vary on;
	gzip_proxied any;
	gzip_comp_level 6;
	gzip_buffers 16 8k;
	gzip_http_version 1.1; ###
	gzip_types text/plain text/css application/json image/svg+xml image/png image/gif image/jpeg application/x-javascript text/xml application/xml application/xml+rss text/javascript font/ttf font/otf font/eot x-font/woff application/x-font-ttf application/x-font-truetype application/x-font-opentype application/font-woff application/font-woff2 application/vnd.ms-fontobject audio/mpeg3 audio/x-mpeg-3 audio/ogg audio/flac audio/mpeg application/mpeg application/mpeg3 application/ogg;
	gzip_min_length 512;

	include /etc/nginx/mime.types;
	include /etc/nginx/uwsgi_params;


	index index.php index.html index.htm;
	default_type application/octet-stream;


	upstream gotify {
		server localhost:8085;
	}

	map $http_upgrade $connection_upgrade {
		default upgrade;
		''      close;
	}




        server {
                server_name push.example.com www.push.example.com;
                listen 80; # IPv4
                return 301 https://$server_name$request_uri;
	}

        server {
                server_name push.example.com www.push.example.com;
		listen 443 ssl http2; # SSL 
		listen [::]:443 ssl http2; # IPv6

		ssi on;

                include /etc/nginx/includes/push.example.com/ssl.conf; # SSL configuration

		server_tokens off;
		add_header Strict-Transport-Security "max-age=63072000; includeSubDomains; preload";
		add_header X-Content-Type-Options "nosniff";
		add_header X-Frame-Options SAMEORIGIN;
		add_header X-XSS-Protection "1; mode=block";
		add_header X-Robots-Tag "none";
		add_header X-Dns-Prefetch-Control 'content=on';
		add_header Expect-CT "enforce,max-age=30";


                location @fallback {

                        proxy_pass http://localhost:8085;

			proxy_set_header Upgrade $http_upgrade;
			proxy_set_header X-Forwarded-Server $host;
			proxy_set_header Connection 'upgrade';
			proxy_set_header Host $host;
			proxy_set_header X-NginX-Proxy true;
			proxy_set_header X-Forwarded-Host $host;
			proxy_set_header X-Real-IP  $remote_addr;
			proxy_set_header X-Forwarded-For $remote_addr;
			proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
			proxy_set_header X-Forwarded-Proto https;
			proxy_set_header X-Forwarded-Port 443;

			proxy_connect_timeout 7d;
			proxy_send_timeout 7d;
			proxy_read_timeout 7d;
			proxy_http_version 1.1;
			proxy_socket_keepalive on;
			proxy_buffering off;







                       access_log /var/log/nginx/gotify_access.log;
                       error_log /var/log/nginx/gotify_error.log;

                }

		
		error_page 503 @maintenance;
		location @maintenance {
			rewrite ^(.*)$ /.maintenance.html break;
		}

	}

}

```

</details>

**Comment by 4oo4** [comment]
I figured this out, this is actually a stupid error on my part. I have localhost mapped to a few different IPs for docker containers, so when I tell it to do a `proxy_pass` to localhost, it's actually trying all of those different IPs instead of just 127.0.0.1.

TL;DR: Read your nginx error logs, and don't get burned by your `/etc/hosts` file :laughing: 

```
2020/03/29 12:09:23 [error] 17164#17164: *303278 connect() failed (113: No route to host) while connecting to upstream, client: 127.0.0.1, server: push.example.com, request: "GET / HTTP/1.1", upstream: "http://172.17.0.2:8085/", host: "push.example.com"
2020/03/29 12:10:26 [error] 15069#15069: *303408 connect() failed (113: No route to host) while connecting to upstream, client: 127.0.0.1, server: push.example.com, request: "GET / HTTP/1.1", upstream: "http://172.17.0.2:8085/", host: "push.example.com"
2020/03/29 12:11:23 [error] 15068#15068: *303489 connect() failed (113: No route to host) while connecting to upstream, client: 127.0.0.1, server: push.example.com, request: "GET / HTTP/1.1", upstream: "http://172.17.0.2:8085/", host: "push.example.com"
```
